### PR TITLE
Request streaming for Apache HttpClient

### PR DESCRIPTION
**Pull request by poutsma**
- Added org.springframework.http.StreamingHttpOutputMessage, which
  allows for a settable request body (as opposed to an output stream).
- Added http.client.HttpComponentsStreamingClientHttpRequest, which
  implements the above mentioned interface, mapping setBody() to a
  setEntity() call on the Apache HttpClient HttpEntityEnclosingRequest.
- Added a 'bufferRequestBody' property to the
  HttpComponentsClientHttpRequestFactory. When this property is set to
  false (default is true), we return a
  HttpComponentsStreamingClientHttpRequest instead of a (request
  buffering) HttpComponentsClientHttpRequest.

Issue: SPR-10728

### CODE REVIEW

File: spring-web/src/main/java/org/springframework/http/client/HttpComponentsStreamingClientHttpRequest.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+	@Override
+	public void setBody(Body body) {
+		assertNotExecuted();
+		this.body = body;
+	}
+
+	@Override
+	protected OutputStream getBodyInternal(HttpHeaders headers) throws IOException {
+		throw new UnsupportedOperationException(
+				"getBody not supported when bufferRequestBody is false");
```
Comment (dreambrother) [inline spring-web/src/main/java/org/springframework/http/client/HttpComponentsStreamingClientHttpRequest.java]: +1
I've tested FormHttpMessageConverter and it is broken for HttpComponentsClientHttpRequestFactory with setBufferRequestBody set to false

Comment (bclozel) [inline spring-web/src/main/java/org/springframework/http/client/HttpComponentsStreamingClientHttpRequest.java]: Could you create an issue on https://jira.spring.io and explain the problem there?
Thanks!

Comment (dreambrother) [inline spring-web/src/main/java/org/springframework/http/client/HttpComponentsStreamingClientHttpRequest.java]: It's already created
https://jira.spring.io/browse/SPR-11199

Comment (bclozel) [inline spring-web/src/main/java/org/springframework/http/client/HttpComponentsStreamingClientHttpRequest.java]: ok thanks!

**Pull request closed without merging**